File: .codex/skills/tech-news-daily/scripts/generate_message.mjs
#!/usr/bin/env node
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";

const SITE_URL = "https://kaistbiztech.github.io/dailytechbrief/";

function parseArgs(argv) {
  const out = { jsonPath: null, projectRoot: null };
  for (let i = 0; i < argv.length; i += 1) {
    if (argv[i] === "--project-root") {
      out.projectRoot = argv[i + 1];
      i += 1;
    } else if (!out.jsonPath) {
      out.jsonPath = argv[i];
    }
  }
  if (!out.jsonPath) throw new Error("Usage: node generate_message.mjs data/YYYY-MM-DD.json [--project-root <root>]");
  return out;
}

function isProjectRoot(dir) {
  return fs.existsSync(path.join(dir, "index.html")) && fs.existsSync(path.join(dir, "data"));
}

function findProjectRoot(jsonPath, explicit) {
  const starts = [explicit, process.env.DAILYTECHBRIEF_ROOT, jsonPath, process.cwd()].filter(Boolean);
  for (const start of starts) {
    let cur = path.resolve(start);
    if (fs.existsSync(cur) && fs.statSync(cur).isFile()) cur = path.dirname(cur);
    while (true) {
      if (isProjectRoot(cur)) return cur;
      const parent = path.dirname(cur);
      if (parent === cur) break;
      cur = parent;
    }
  }
  throw new Error("Project root not found. Run from the repo or pass --project-root.");
}

function escText(value) {
  return String(value ?? "");
}

function buildText(edition) {
  const lines = [
    "📌 KAIST 경영대학 테크 네트워크",
    "데일리 테크 브리프",
    `${edition.id.replaceAll("-", ".")} ${edition.dayOfWeek}요일`,
    "",
    `전체 보기 👉 ${SITE_URL}date/${edition.id}/`,
    "",
    "🔎 오늘의 키워드",
    edition.newsItems.flatMap((item) => item.keywords ?? []).join(" · "),
    "",
  ];

  for (const item of edition.newsItems) {
    lines.push(`${String(item.order).padStart(2, "0")}. ${item.title}`);
    lines.push(item.summary);
    lines.push("");
  }

  lines.push(`전체 보기 👉 ${SITE_URL}date/${edition.id}/`);
  lines.push("KAIST 경영대학 테크 네트워크");
  return lines.join(os.EOL);
}

function psString(value) {
  return `'${String(value).replaceAll("'", "''")}'`;
}

function renderWithPowerShell(edition, projectRoot, cardPath, ogPath) {
  if (process.platform !== "win32") {
    throw new Error("PNG fallback renderer requires Windows PowerShell. Use the Python Playwright script on non-Windows hosts.");
  }

  const headlines = edition.newsItems.map((item) => item.title);
  const keywords = edition.newsItems.flatMap((item) => item.keywords ?? []);
  const script = `
$ErrorActionPreference = 'Stop'
Add-Type -AssemblyName System.Drawing
$projectRoot = ${psString(projectRoot)}
$cardPath = ${psString(cardPath)}
$ogPath = ${psString(ogPath)}
$dateLine = ${psString(`${edition.id.replaceAll("-", ".")} ${edition.dayOfWeek}요일`)}
$headlines = @(${headlines.map(psString).join(",")})
$keywords = @(${keywords.map(psString).join(",")})
$logoPath = Join-Path $projectRoot 'KCB_Logo.png'

function New-Font([string]$Family, [float]$Size, [string]$Style = 'Bold') {
  New-Object System.Drawing.Font($Family, $Size, [System.Drawing.FontStyle]::$Style, [System.Drawing.GraphicsUnit]::Pixel)
}

function Draw-Edition([int]$Width, [int]$Height, [string]$OutPath, [bool]$Og) {
  $bmp = New-Object System.Drawing.Bitmap $Width, $Height
  $g = [System.Drawing.Graphics]::FromImage($bmp)
  $g.SmoothingMode = [System.Drawing.Drawing2D.SmoothingMode]::AntiAlias
  $g.TextRenderingHint = [System.Drawing.Text.TextRenderingHint]::ClearTypeGridFit
  $bg = [System.Drawing.Color]::FromArgb(31,72,153)
  $deep = [System.Drawing.Color]::FromArgb(21,50,107)
  $g.Clear($bg)
  $logo = [System.Drawing.Image]::FromFile($logoPath)
  $white = [System.Drawing.Brushes]::White
  $muted = New-Object System.Drawing.SolidBrush ([System.Drawing.Color]::FromArgb(210,225,255))
  $deepBrush = New-Object System.Drawing.SolidBrush $deep

  if ($Og) {
    $g.FillRectangle($deepBrush, 40, 260, 1120, 260)
    $g.DrawImage($logo, 56, 44, 250, 74)
    $g.DrawString('KAIST Business Tech Network', (New-Font 'Segoe UI' 34), $white, 330, 58)
    $g.DrawString('Daily Tech Brief', (New-Font 'Segoe UI' 68), $white, 56, 140)
    $g.DrawString($dateLine, (New-Font 'Malgun Gothic' 42), $white, 760, 150)
    $g.DrawString('Keywords', (New-Font 'Segoe UI' 20), $white, 72, 286)
    $font = New-Font 'Malgun Gothic' 28
    $x = 72; $y = 336
    foreach ($kw in $keywords[0..([Math]::Min($keywords.Length - 1, 9))]) {
      $size = $g.MeasureString($kw, $font)
      if ($x + $size.Width -gt 1080) { $x = 72; $y += 48 }
      $g.DrawString($kw, $font, $white, $x, $y)
      $x += [int]$size.Width + 24
    }
  } else {
    $g.FillEllipse((New-Object System.Drawing.SolidBrush ([System.Drawing.Color]::FromArgb(35,255,255,255))), 700, -120, 520, 520)
    $g.FillRectangle($deepBrush, 70, 360, 940, 250)
    $g.DrawImage($logo, 70, 70, 270, 86)
    $g.DrawString('KAIST Business Tech Network', (New-Font 'Segoe UI' 34), $white, 370, 88)
    $g.DrawString('Daily Tech Brief', (New-Font 'Segoe UI' 64), $white, 70, 190)
    $g.DrawString($dateLine, (New-Font 'Malgun Gothic' 52), $white, 70, 270)
    $kwFont = New-Font 'Malgun Gothic' 24
    $x = 100; $y = 398
    foreach ($kw in $keywords) {
      $size = $g.MeasureString($kw, $kwFont)
      if ($x + $size.Width -gt 950) { $x = 100; $y += 36 }
      if ($y -lt 590) { $g.DrawString($kw, $kwFont, $white, $x, $y) }
      $x += [int]$size.Width + 18
    }
    $numFont = New-Font 'Consolas' 26
    $headFont = New-Font 'Malgun Gothic' 28
    $y = 655
    for ($i = 0; $i -lt $headlines.Length; $i++) {
      $g.DrawString(('{0:D2}' -f ($i + 1)), $numFont, $muted, 80, $y + 4)
      $g.DrawString($headlines[$i], $headFont, $white, 150, $y)
      $y += 92
    }
    $g.FillRectangle([System.Drawing.Brushes]::White, 760, 1780, 250, 70)
    $g.DrawString('View All ->', (New-Font 'Segoe UI' 28), (New-Object System.Drawing.SolidBrush $bg), 808, 1798)
  }

  $bmp.Save($OutPath, [System.Drawing.Imaging.ImageFormat]::Png)
  $logo.Dispose(); $deepBrush.Dispose(); $muted.Dispose(); $g.Dispose(); $bmp.Dispose()
}

Draw-Edition 1080 1920 $cardPath $false
Draw-Edition 1200 630 $ogPath $true
`;

  const result = spawnSync("powershell.exe", ["-NoProfile", "-ExecutionPolicy", "Bypass", "-Command", script], {
    encoding: "utf8",
  });
  if (result.status !== 0) {
    throw new Error(result.stderr || result.stdout || "PowerShell renderer failed.");
  }
}

const args = parseArgs(process.argv.slice(2));
const projectRoot = findProjectRoot(args.jsonPath, args.projectRoot);
const jsonPath = path.isAbsolute(args.jsonPath) ? args.jsonPath : path.resolve(process.cwd(), args.jsonPath);
const edition = JSON.parse(fs.readFileSync(jsonPath, "utf8"));

const messageDir = path.join(projectRoot, "Message", edition.id);
const dateDir = path.join(projectRoot, "date", edition.id);
fs.mkdirSync(messageDir, { recursive: true });
fs.mkdirSync(dateDir, { recursive: true });

const textPath = path.join(messageDir, "text.txt");
const cardPath = path.join(messageDir, "card.png");
const ogPath = path.join(dateDir, "og.png");

fs.writeFileSync(textPath, buildText(edition), "utf8");
console.log(`[OK] wrote ${path.relative(projectRoot, textPath)}`);
renderWithPowerShell(edition, projectRoot, cardPath, ogPath);
console.log(`[OK] wrote ${path.relative(projectRoot, cardPath)}`);
console.log(`[OK] wrote ${path.relative(projectRoot, ogPath)}`);

File: .codex/skills/tech-news-daily/scripts/generate_message.py
#!/usr/bin/env python3
"""
Daily Tech Brief - 카톡 공유 산출물 생성기

입력: data/YYYY-MM-DD.json
출력:
  - Message/YYYY-MM-DD/card.png   (로컬 카톡 첨부 사본)
  - Message/YYYY-MM-DD/text.txt   (로컬 카톡 복붙 텍스트)
  - date/YYYY-MM-DD/og.png        (사이트 OG 이미지, 깃 푸시)

OG 메타가 박힌 일자별 정적 HTML은 build_site.py가 생성한다.

사용법:
    python generate_message.py data/2026-05-26.json [--project-root <dailytechbrief-root>]
"""
import argparse
import json
import os
import shutil
import sys
from pathlib import Path

SITE_URL = "https://kaistbiztech.github.io/dailytechbrief/"
PROJECT_ROOT = None
TEMPLATES_DIR = Path(__file__).resolve().parents[1] / "templates"
KAKAO_TEMPLATE = TEMPLATES_DIR / "kakao-card.html"   # 9:16 세로, 카톡 앱 첨부용
OG_TEMPLATE = TEMPLATES_DIR / "og-card.html"          # 1.91:1 가로, 메신저 OG 미리보기용
MESSAGE_BASE = None
DATE_BASE = None
LOGO_PATH = None


def _is_project_root(path: Path) -> bool:
    return (path / "index.html").is_file() and (path / "data").is_dir()


def _walk_for_project_root(start: Path) -> Path | None:
    start = start.resolve()
    candidates = [start] + list(start.parents)
    for candidate in candidates:
        if _is_project_root(candidate):
            return candidate
    return None


def resolve_project_root(json_arg: Path, explicit: str | None = None) -> Path:
    candidates = []
    if explicit:
        candidates.append(Path(explicit))
    if os.environ.get("DAILYTECHBRIEF_ROOT"):
        candidates.append(Path(os.environ["DAILYTECHBRIEF_ROOT"]))
    candidates.extend([json_arg, Path.cwd()])

    for candidate in candidates:
        found = _walk_for_project_root(candidate)
        if found:
            return found

    raise SystemExit(
        "Project root not found. Run from the dailytechbrief repo or pass --project-root."
    )


def configure_paths(project_root: Path) -> None:
    global PROJECT_ROOT, MESSAGE_BASE, DATE_BASE, LOGO_PATH
    PROJECT_ROOT = project_root
    MESSAGE_BASE = PROJECT_ROOT / "Message"
    DATE_BASE = PROJECT_ROOT / "date"
    LOGO_PATH = PROJECT_ROOT / "KCB_Logo.png"


def build_text(edition: dict) -> str:
    """카톡 복붙용 텍스트 — 사이트 링크 상단, KAIST 프레임, 키워드, 3문장 요약."""
    date_str = edition["id"].replace("-", ".")
    dow = edition.get("dayOfWeek", "")
    eid = edition["id"]

    parts = []
    parts.append("📰 KAIST 경영대학 테크 네트워크")
    parts.append("데일리 테크 브리프")
    parts.append(f"{date_str} {dow}요일")
    parts.append("")
    parts.append(f"전체 보기 👉 {SITE_URL}date/{eid}/")

    # 키워드 묶음
    all_kw = []
    for it in edition["newsItems"]:
        all_kw.extend(it.get("keywords", []))
    if all_kw:
        parts.append("")
        parts.append("🔑 오늘의 키워드")
        parts.append(" · ".join(all_kw))

    parts.append("")
    parts.append("━━━━━━━━━━━━━━━━━━━")
    parts.append("")

    for item in edition["newsItems"]:
        parts.append(f"{item['order']:02d}. {item['title']}")
        parts.append(item["summary"])
        parts.append("")

    parts.append("━━━━━━━━━━━━━━━━━━━")
    parts.append(f"전체 보기 👉 {SITE_URL}date/{eid}/")
    parts.append("© KAIST 경영대학 테크 네트워크")
    return "\n".join(parts)


def _logo_data_url() -> str | None:
    import base64
    if not LOGO_PATH.is_file():
        return None
    return "data:image/png;base64," + base64.b64encode(LOGO_PATH.read_bytes()).decode("ascii")


def _capture(browser, template_path: Path, edition: dict, viewport: dict, out_path: Path, logo_url: str | None) -> None:
    ctx = browser.new_context(viewport=viewport, device_scale_factor=2)
    page = ctx.new_page()
    page.goto(template_path.resolve().as_uri(), wait_until="networkidle")
    page.evaluate(
        "([ed, logo]) => window.renderEdition(ed, logo)",
        [edition, logo_url],
    )
    page.wait_for_timeout(700)
    out_path.parent.mkdir(parents=True, exist_ok=True)
    page.screenshot(path=str(out_path), full_page=False, type="png")
    ctx.close()


def generate_cards(edition: dict, kakao_out: Path, og_out: Path) -> None:
    """
    카톡 앱 첨부용 9:16 세로 카드 + 메신저 OG 1.91:1 가로 카드를 각각 캡처.

    - kakao_out: 1080×1920 (Message/<id>/card.png)
    - og_out: 1200×630 (date/<id>/og.png)
    """
    from playwright.sync_api import sync_playwright

    logo_url = _logo_data_url()
    with sync_playwright() as p:
        browser = p.chromium.launch()
        _capture(browser, KAKAO_TEMPLATE, edition,
                 {"width": 1080, "height": 1920}, kakao_out, logo_url)
        _capture(browser, OG_TEMPLATE, edition,
                 {"width": 1200, "height": 630}, og_out, logo_url)
        browser.close()


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("json_path", help="path to data/YYYY-MM-DD.json")
    parser.add_argument("--project-root", help="dailytechbrief repository root")
    args = parser.parse_args()

    json_path = Path(args.json_path)
    project_root = resolve_project_root(json_path, args.project_root)
    configure_paths(project_root)
    if not json_path.is_absolute():
        candidate = (Path.cwd() / json_path).resolve()
        if not candidate.is_file():
            candidate = (project_root / json_path).resolve()
        json_path = candidate
    else:
        json_path = json_path.resolve()

    if not json_path.is_file():
        print(f"Not found: {json_path}", file=sys.stderr)
        return 1

    edition = json.loads(json_path.read_text(encoding="utf-8"))
    eid = edition["id"]

    message_dir = MESSAGE_BASE / eid
    message_dir.mkdir(parents=True, exist_ok=True)
    date_dir = DATE_BASE / eid
    date_dir.mkdir(parents=True, exist_ok=True)

    # 1) 텍스트 (로컬)
    text_path = message_dir / "text.txt"
    text_path.write_text(build_text(edition), encoding="utf-8")
    print(f"[OK] wrote {text_path.relative_to(PROJECT_ROOT)}")

    # 2) 카드 PNG: 카톡용(9:16) + OG용(1.91:1) 각각 캡처
    card_path = message_dir / "card.png"
    og_path = date_dir / "og.png"
    generate_cards(edition, kakao_out=card_path, og_out=og_path)
    print(f"[OK] wrote {card_path.relative_to(PROJECT_ROOT)}")
    print(f"[OK] wrote {og_path.relative_to(PROJECT_ROOT)}")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
